File: hardsploit/modules/i2c/i2c_sniffer.rb
#!/usr/bin/ruby
#===================================================
#  Hardsploit API - By Opale Security
#  www.opale-security.com || www.hardsploit.io
#  License: GNU General Public License v3
#  License URI: http://www.gnu.org/licenses/gpl.txt
#===================================================

require_relative '../../Core/HardsploitAPI'
class HardsploitAPI_I2C_SNIFFER
public

	def initialize(timeout:)
		#to be sure the singleton was initialize
		HardsploitAPI.instance.connect
		self.timeout=timeout
		#i2c_SetSettings # fonction ? class ?
	end

	def timeout
		return @timeout
	end

	def timeout=(timeout)
		@timeout = timeout
	end

	def i2c_start_sniff(path:)
		packet = HardsploitAPI.prepare_packet # 0x50 Command raw communication to FPGA fifo included inside prepare_packet
		packet.push 0x55 #Command start sniffer
		#First we will try to have the first frame only, and then all frames
		HardsploitAPI.instance.sendPacket(packet)
		file = File.open(path,"wb")
		stop = FALSE
		while (stop == FALSE)
			begin
				result = HardsploitAPI.instance.receiveDATA(@timeout).bytes.to_a.drop(4)#.drop(4) #5s = time out car sniffer, assez aléatoire pour savoir quand une donnée va arriver
				p result
				file.write result.pack('C*')
			rescue LIBUSB::ERROR_TIMEOUT
				end_packet = HardsploitAPI.prepare_packet # 0x50 Command raw communication to FPGA fifo include inside prepare_packet
				end_packet.push 0xaa #Command stop sniffer
				begin
					result = HardsploitAPI.instance.sendAndReceiveDATA(end_packet,1000).drop(4) #timeout 1s
					p result
					file.write result.pack('C*')
					file.close
				end
				stop = TRUE
			end
			sleep(0.01)
		end
		p result.size
		return result
	end

	def i2c_stop_sniff
		end_packet = HardsploitAPI.prepare_packet # 0x50 Command raw communication to FPGA fifo include inside prepare_packet
		end_packet.push 0xaa #Command stop sniffer
		begin
			p result = HardsploitAPI.instance.sendAndReceiveDATA(end_packet,1000).drop(4) #timeout 1s
			p result.size
			#HardsploitAPI.instance.sendPacket(end_packet)	#We send the stop command and we wait for the last data which are not already transmitted
																						#What can we do if there is no data...
	#	rescue
	#		raise HardsploitAPI::ERROR::USB_ERROR
		end
		return result
	end
end

File: hardsploit/modules/i2c/i2c_sniffer.py
import time

import usb.core


class HardsploitI2cSniffer:

    def __init__(self, timeout, hardsploit_api):
        self._timeout = timeout
        self._api = hardsploit_api

    @property
    def timeout(self):
        return self._timeout

    @timeout.setter
    def timeout(self, timeout):
        self._timeout = timeout

    def i2c_start_sniff(self, path):
        packet = self._api.prepare_packet()
        packet.append(0x55)
        self._api.send_packet(packet)
        result = ""
        file = open(path, "wb")
        stop = False
        while not stop:
            try:
                result = list(self._api.receive_data(self.timeout))[4:]
                print(result)
                file.write(bytes(result))

            except usb.core.USBTimeoutError:
                end_packet = self._api.prepare_packet()
                end_packet.append(0xaa)
                try:
                    result = list(self._api.send_and_receive_data(end_packet, 1000))[4:]
                    print(result)
                    file.write(bytes(result))
                    file.close()

                except Exception:
                    pass

                stop = True
            time.sleep(0.01)

        print(len(result))
        return result

    def i2c_stop_sniff(self):
        end_packet = self._api.prepare_packet()
        end_packet.append(0xaa)
        result = ""
        try:
            result = list(self._api.send_and_receive_data(end_packet, 1000))[4:]
            print(result)
            print(len(result))
        except Exception:
            pass

        return result
